formal-spec/Leios/Trace/Verifier.agda

```diff
@@ -0,0 +1,439 @@
+open import Leios.Prelude hiding (id)
+
+module Leios.Trace.Verifier (numberOfParties : ℕ) (SUT-id : Fin numberOfParties) where
+
+open import Leios.SpecStructure using (SpecStructure)
+open import Leios.Defaults numberOfParties SUT-id using (st; sd; LeiosState; initLeiosState; isb; hpe; hhs; htx; SendIB; FFDState; Dec-SimpleFFD; send-total; fetch-total)
+open SpecStructure st hiding (Hashable-IBHeader; Hashable-EndorserBlock; isVoteCertified)
+
+open import Leios.Short st hiding (LeiosState; initLeiosState)
+open import Prelude.Closures _↝_
+open GenFFD
+
+data FFDUpdate : Type where
+  IB-Recv-Update : InputBlock → FFDUpdate
+  EB-Recv-Update : EndorserBlock → FFDUpdate
+  VT-Recv-Update : List Vote → FFDUpdate
+
+data Action : Type where
+  IB-Role-Action : ℕ → Action
+  EB-Role-Action : ℕ → List String → Action
+  VT-Role-Action : ℕ → Action
+  No-IB-Role-Action No-EB-Role-Action No-VT-Role-Action : Action
+  Ftch-Action : Action
+  Slot-Action : ℕ → Action
+  Base₁-Action : Action
+  Base₂a-Action : EndorserBlock → Action
+  Base₂b-Action : Action
+
+Actions = List (Action × LeiosInput)
+
+private variable
+  s s′ : LeiosState
+  α : Action
+
+data ValidUpdate : FFDUpdate → LeiosState → Type where
+
+  IB-Recv : ∀ {ib} →
+    ValidUpdate (IB-Recv-Update ib) s
+
+  EB-Recv : ∀ {eb} →
+    ValidUpdate (EB-Recv-Update eb) s
+
+  VT-Recv : ∀ {vt} →
+    ValidUpdate (VT-Recv-Update vt) s
+
+data ValidAction : Action → LeiosState → LeiosInput → Type where
```

Comment:
Why are some of the premises marked irrelevant but not all of them? Is this irrelevance for performance reasons?

Reply:
Marking premises irrelevant made some proofs easier. In case a premise is need, it is recomputed